src/ripple/app/ledger/impl/LedgerMaster.cpp

```diff
@@ -1037,6 +1039,51 @@ LedgerMaster::checkAccept(std::shared_ptr<Ledger const> const& ledger)
     app_.getFeeTrack().setRemoteFee(fee);
 
     tryAdvance();
+
+    if (ledger->seq() % 256 == 0)
+    {
+        /*
+         * Check if the majority of validators run a higher version rippled
+         * software. If so print a warning.
+         *
+         * Once the HardenedValidations amendment is enabled, validators include
+         * their rippled software version in the validation messages of every
+         * (flag - 1) ledger. We wait for one ledger time before checking the
+         * version information to accumulate more validation messages.
+         */
+        auto const& patentHash = ledger->info().parentHash;
+        auto const parentLedger = getLedgerByHash(patentHash);
+        if (parentLedger &&
+            parentLedger->rules().enabled(featureHardenedValidations))
+        {
+            auto const vals =
+                app_.getValidations().getTrustedForLedger(patentHash);
+            auto higherVersionCount = std::count_if(
+                vals.begin(), vals.end(), [](auto const& v) -> bool {
+                    if (v->isFieldPresent(sfServerVersion))
+                        return BuildInfo::localVersionLow(
+                            v->getFieldU64(sfServerVersion));
+                    else
+                        return false;
+                });
+            auto const threshold = static_cast<std::size_t>(std::ceil(
+                app_.validators().getQuorumKeys().second.size() *
+                BuildInfo::versionUpgradeWarningThreshold));
+            if (higherVersionCount >= threshold)
+            {
+                std::cerr << "***********************************************"
```

Comment:
A single statement will do:

```
        std::cerr << "Check for upgrade: "
                     "A majority of trusted validators are running a newer "
                     "version.\n";
```

Reply:
Thanks, changed.

Reply:
I think it would be useful to both write to `cerr` and log (at error level maybe?), in case of systems that _ignore_ `std::cerr`. Not everybody goes through the log either, but that would increase the chances of catching it. And since you mentioned doing it once a week, it won't be a big deal.

Reply:
I think that is fine. @nbougalis what do you think?

Reply:
That would be great.